cmd/minikube/cmd/dashboard.go

```diff
@@ -17,68 +17,136 @@ limitations under the License.
-	Long:  `Opens/displays the kubernetes dashboard URL for your local cluster`,
+	Short: "Access the kubernetes dashboard running within the minikube cluster",
+	Long:  `Access the kubernetes dashboard running within the minikube cluster`,
 	Run: func(cmd *cobra.Command, args []string) {
 		api, err := machine.NewAPIClient()
+		defer func() {
+			err := api.Close()
+			if err != nil {
+				glog.Warningf("Failed to close API: %v", err)
+			}
+		}()
+
 		if err != nil {
-			fmt.Fprintf(os.Stderr, "Error getting client: %s\n", err)
+			fmt.Fprintf(os.Stderr, "Error creating client: %v\n", err)
 			os.Exit(1)
 		}
-		defer api.Close()
-
 		cluster.EnsureMinikubeRunningOrExit(api, 1)
-		namespace := "kube-system"
-		svc := "kubernetes-dashboard"
 
-		if err = commonutil.RetryAfter(20, func() error { return service.CheckService(namespace, svc) }, 6*time.Second); err != nil {
-			fmt.Fprintf(os.Stderr, "Could not find finalized endpoint being pointed to by %s: %s\n", svc, err)
+		ns := "kube-system"
+		svc := "kubernetes-dashboard"
+		if err = util.RetryAfter(30, func() error { return service.CheckService(ns, svc) }, 1*time.Second); err != nil {
+			fmt.Fprintf(os.Stderr, "%s:%s is not running: %v\n", ns, svc, err)
 			os.Exit(1)
 		}
 
-		urls, err := service.GetServiceURLsForService(api, namespace, svc, template.Must(template.New("dashboardServiceFormat").Parse(defaultServiceFormatTemplate)))
+		p, hostPort, err := kubectlProxy()
 		if err != nil {
-			fmt.Fprintln(os.Stderr, err)
-			fmt.Fprintln(os.Stderr, "Check that minikube is running.")
-			os.Exit(1)
+			glog.Fatalf("kubectl proxy: %v", err)
 		}
-		if len(urls) == 0 {
-			errMsg := "There appears to be no url associated with dashboard, this is not expected, exiting"
-			glog.Infoln(errMsg)
+		url := dashboardURL(hostPort, ns, svc)
+
+		if err = util.RetryAfter(60, func() error { return checkURL(url) }, 1*time.Second); err != nil {
+			fmt.Fprintf(os.Stderr, "%s is not responding properly: %v\n", url, err)
 			os.Exit(1)
 		}
+
 		if dashboardURLMode {
-			fmt.Fprintln(os.Stdout, urls[0])
+			fmt.Fprintln(os.Stdout, url)
 		} else {
-			fmt.Fprintln(os.Stdout, "Opening kubernetes dashboard in default browser...")
-			browser.OpenURL(urls[0])
+			fmt.Fprintln(os.Stdout, fmt.Sprintf("Opening %s in your default browser...", url))
+			if err = browser.OpenURL(url); err != nil {
+				fmt.Fprintf(os.Stderr, fmt.Sprintf("failed to open browser: %v", err))
+			}
+		}
+
+		glog.Infof("Waiting forever for kubectl proxy to exit ...")
+		if err = p.Wait(); err != nil {
+			glog.Errorf("Wait: %v", err)
 		}
 	},
 }
 
+// kubectlProxy runs "kubectl proxy", returning host:port
+func kubectlProxy() (*exec.Cmd, string, error) {
+	path, err := exec.LookPath("kubectl")
+	if err != nil {
+		return nil, "", errors.Wrap(err, "kubectl not found in PATH")
+	}
+
+	// port=0 picks a random system port.
+	cmd := exec.Command(path, "proxy", "--port=0")
```

Comment:
Is there a reason we start a subprocess here ? Don't we have all the mechanics to start a server in-process ?

We have an in-house implementation that does this, and we like that we don't mess with subprocess lifetime management (which is harder on Windows), and orphaned executables:

```
func getLocalConfig() (*api.Config, error) {
	clientConfig, err := k8s.ReadConfigOrNew(k8s.GetKubeConfigPath())
	clientConfig.CurrentContext = context // by default: minikube
	return clientConfig, err
}

func startProxy(proxyPort int, context string) (net.Listener, error) {
	// This is strongly inspired by https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/proxy.go,
	// the kubectl proxy implementation. We use this implementation style to limit subprocess interaction.
	// We hardcode values for our specific needs.
	// The equivalent is: $ kubectl --context=minikube proxy --accept-hosts='.*' --address='0.0.0.0'

	// default filter
	filter := &proxy.FilterServer{
		AcceptPaths:   proxy.MakeRegexpArrayOrDie(proxy.DefaultPathAcceptRE),
		RejectPaths:   proxy.MakeRegexpArrayOrDie(proxy.DefaultPathRejectRE),
		AcceptHosts:   proxy.MakeRegexpArrayOrDie(".*"),
		RejectMethods: proxy.MakeRegexpArrayOrDie(proxy.DefaultMethodRejectRE),
	}

	var l net.Listener
	// Get minikube context which will be forwarded. The user must have this context in his ~/.kube/config.
	contextConfig, err := clientcmd.BuildConfigFromKubeconfigGetter("", clientcmd.KubeconfigGetter(getLocalConfig))
	if err != nil {
		return l, err
	}

	proxyServer, err := proxy.NewServer("", "/", "/static/", filter, contextConfig)

	// Separate listening from serving so we can report the bound port
	// when it is chosen by os (eg: port == 0)
	l, err = proxyServer.Listen("0.0.0.0", proxyPort)
	if err != nil {
		return l, err
	}

	fmt.Printf("Proxying to \"%s\" on %s\n", context, l.Addr().String())
	go proxyServer.ServeOnListener(l)

	return l, nil
}
```

Reply:
> Is there a reason we start a subprocess here ? Don't we have all the mechanics to start a server in-process?

We do, but not the code. After discussing it offline with @balopat, we decided to move forward with executing kubectl in this iteration. The primary worry for me was the maintenance burden of vendoring in the proxy and keeping up with future API changes. The command-line input/output interface seems more likely to be robust.

I do agree that zombies might be an issue: if they seem to be so, I'll happily move toward vendoring the proxy server in a future PR.

Reply:
I agree that this can be a slightly better user experience and can be a useful follow-up feature after this PR is merged, especially if there will be more evidence around people having issues with the subprocess approach, but I think we can move forward with the shell-out approach for now.